plugin/evm/atomic_trie_test.go

```diff
@@ -591,6 +597,187 @@ func TestApplyToSharedMemory(t *testing.T) {
+			wantHasCommitted:        true,
+			wantLastCommittedHeight: 110,
+			wantLastCommittedRoot:   common.Hash{3},
+			wantLastAcceptedRoot:    common.Hash{3},
+			wantTipBufferRoot:       common.Hash{3},
+			wantKeyValuePairs: map[string]string{
+				metadataHexPrefix + "0000000000000064":// height 100
+				"0200000000000000000000000000000000000000000000000000000000000000",
+				metadataHexPrefix + "000000000000006e":// height 110
+				"0300000000000000000000000000000000000000000000000000000000000000",
+				metadataHexPrefix + hex.EncodeToString(lastCommittedKey): "000000000000006e", // height 110
+			},
+		},
+	}
+
+	for name, testCase := range testCases {
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			versionDB := versiondb.New(memdb.New())
+			atomicTrieDB := prefixdb.New(atomicTrieDBPrefix, versionDB)
+			metadataDB := prefixdb.New(atomicTrieMetaDBPrefix, versionDB)
+			const lastAcceptedHeight = 0 // no effect
+			atomicTrie, err := newAtomicTrie(atomicTrieDB, metadataDB, atomic.TestTxCodec,
+				lastAcceptedHeight, testCase.commitInterval)
+			atomicTrie.lastAcceptedRoot = testCase.lastAcceptedRoot
+			if testCase.lastAcceptedRoot != types.EmptyRootHash {
+				// Generate trie node test blob
+				encoder := rlp.NewEncoderBuffer(nil)
+				offset := encoder.List()
+				encoder.WriteBytes([]byte{1})        // key
+				encoder.WriteBytes(make([]byte, 32)) // value
+				encoder.ListEnd(offset)
+				testBlob := encoder.ToBytes()
+				err := encoder.Flush()
+				require.NoError(t, err)
+
+				nodeSet := trienode.NewNodeSet(testCase.lastAcceptedRoot)
+				nodeSet.AddNode([]byte("any"), trienode.New(testCase.lastAcceptedRoot, testBlob)) // dirty node
+				err = atomicTrie.InsertTrie(nodeSet, testCase.lastAcceptedRoot)
+				require.NoError(t, err)
+
+				_, storageSize, _ := atomicTrie.trieDB.Size()
+				require.NotZero(t, storageSize, "there should be a dirty node taking up storage space")
+			}
+			atomicTrie.updateLastCommitted(testCase.lastCommittedRoot, testCase.lastCommittedHeight)
+			require.NoError(t, err)
+
+			hasCommitted, err := atomicTrie.AcceptTrie(testCase.height, testCase.root)
+			if testCase.wantErr == "" {
+				require.NoError(t, err)
+			} else {
+				assert.EqualError(t, err, testCase.wantErr)
+			}
+
+			assert.Equal(t, testCase.wantHasCommitted, hasCommitted)
+			assert.Equal(t, testCase.wantLastCommittedHeight, atomicTrie.lastCommittedHeight)
+			assert.Equal(t, testCase.wantLastCommittedRoot, atomicTrie.lastCommittedRoot)
+			assert.Equal(t, testCase.wantLastAcceptedRoot, atomicTrie.lastAcceptedRoot)
+
+			// Check dereferencing previous dirty root inserted occurred
+			_, storageSize, _ := atomicTrie.trieDB.Size()
+			assert.Zerof(t, storageSize, "storage size should be zero after accepting the trie due to the dirty nodes derefencing but is %s", storageSize)
+
+			keyValuePairs := make(map[string]string, len(testCase.wantKeyValuePairs))
+			it := versionDB.NewIterator()
+			for it.Next() {
```

Comment:
The iterator should be released.
Additionally when it.Next() returns false, the caller should check it.Error() to see if iteration was terminated early due to an error

Not sure whether it's necessary for an in memory db, but it is the iterator interface